### PR TITLE
fix(indexer)!: add indexed id to fix implicit aliases and nfts

### LINKED ISSUE #1074: Outputs with implicit alias and NFT ids are not indexed correctly
We currently don't handle the implicit ids of alias and NFT ids correctly.

We should create an additional field (with corresponding index) in `output.details` to help with the query.

### PR DESCRIPTION
**Pull request by Alexandcoats**
## Linked Issues

* Closes #1074

## Notes to Reviewer

As a reviewer, please pay particular attention to the following areas when reviewing this PR and tick the above boxes after you have completed the steps.

#### API Changes
* [x] Test the indexer API endpoints.

#### INX Changes
* [x] Run chronicle using an INX connection.

#### Database Changes
* [x] Review database queries for correctness/conciseness.
* [x] Ensure queries are supported by indexes if needed.
* [x] Check for breaking changes in the data model and matching (conventional) commit message prefix.
* [x] Run the migration script.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -129,3 +129,7 @@ inherits = "release"
 codegen-units = 1
 lto = true
 strip = "symbols"
+
+[[bin]]
+name = "1-0-0-beta-31"
+path = "src/migrations/1.0.0-beta.31.rs"
```
Comment (grtlr) [inline Cargo.toml:135]: We should add migrations as a subcommand because it makes it much more convenient to use in our released Docker image.

File: src/migrations/1.0.0-beta.31.rs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    pub mongodb_database_name: String,
+}
+
+#[tokio::main]
+async fn main() -> eyre::Result<()> {
+    let args = MongoDbArgs::parse();
+    let config = MongoDbConfig {
+        conn_str: args.mongodb_conn_str,
+        database_name: args.mongodb_database_name,
+    };
```
Comment (grtlr) [inline src/migrations/1.0.0-beta.31.rs:38]: What's nice about subcommands is that they can reuse the already existing MongoDb args struct, so we could get rid of this code.

File: src/migrations/1.0.0-beta.31.rs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        // Alias and nft are the same length so both can be done this way since they are just serialized as bytes
+        let id = AliasId::from(output_id);
+        collection
+            .update_one(
+                doc! { "$_id": output_id },
+                doc! { "$set": { "details.indexed_id": id } },
+                None,
+            )
+            .await?;
+    }
```
Comment (grtlr) [inline src/migrations/1.0.0-beta.31.rs:77]: Why not use a simple [`updateMany`](https://www.mongodb.com/docs/manual/reference/method/db.collection.updateMany/#mongodb-method-db.collection.updateMany) query?

There is no reason to retrieve all this data from the DB. It also makes the intent much clearer.

File: src/migrations/1.0.0-beta.31.rs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        output_id: OutputId,
+    }
+
+    let outputs = collection
+        .aggregate::<Res>(
+            [
+                doc! { "$match": { "$or": [
+                    { "output.alias_id": AliasId::implicit() },
+                    { "output.nft_id": NftId::implicit() }
+                ] } },
```
Comment (grtlr) [inline src/migrations/1.0.0-beta.31.rs:55]: Aren't we missing the outputs with explicit ideas?

File: src/db/collections/outputs/indexer/mod.rs [hunk 5]
```diff
@@ -75,16 +76,11 @@ impl OutputCollection {
...
-            IndexedId::Alias(_) => "output.alias_id",
-            IndexedId::Foundry(_) => "output.foundry_id",
-            IndexedId::Nft(_) => "output.nft_id",
-        };
         let mut res = self
             .aggregate(
                 [
                     doc! { "$match": {
-                        id_string: id,
+                        "output.details.indexed_id": id,
```
Comment (grtlr) [inline src/db/collections/outputs/indexer/mod.rs:83]: I'm not sure this is correct. The indexer API differentiates between NFTs, aliases, and foundries. As far as I can tell we never check if the output returned from the database matches the queried output type.

File: Cargo.toml (hunk 1 above)
Comment (grtlr) [reply Cargo.toml:135]: I don't think that's worth the updated complexity. The inflation of the binary is probably negligible.

Comment (Alex6323) [reply Cargo.toml:135]: We most definitely will have migrations in the future so a `migrate` subcommand makes sense in any case IMO. The only thing that this subcommand is then supposed to do is to check whether a certain migration is applicable by checking the versions and then execute the migration binary, or deny it with an appropriate message.

Comment (Alex6323) [reply Cargo.toml:135]: It's maybe negligible for the first migration :trollface:

File: src/migrations/1.0.0-beta.31.rs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    )]
+    pub mongodb_conn_str: String,
+    /// The MongoDb database name.
+    #[arg(long, value_name = "NAME", default_value = mongocfg::DEFAULT_DATABASE_NAME)]
+    pub mongodb_database_name: String,
+}
+
+#[tokio::main]
+async fn main() -> eyre::Result<()> {
+    let args = MongoDbArgs::parse();
```
Comment (Alex6323) [inline src/migrations/1.0.0-beta.31.rs:34]: I think you forgot to read the environment variables.

File: src/migrations/1.0.0-beta.31.rs (hunk 4 above)
Comment (Alex6323) [reply src/migrations/1.0.0-beta.31.rs:55]: yup, I agree. I think this `$match` should merely check for existence of those ids and in then make the distinction in the `for` loop below.

File: src/migrations/1.0.0-beta.31.rs (hunk 3 above)
Comment (Alex6323) [reply src/migrations/1.0.0-beta.31.rs:77]: It's just a migration script. I would handle both cases in the same loop with `updateOne` tbh.

File: src/db/collections/outputs/indexer/mod.rs (hunk 5 above)
Comment (Alex6323) [reply src/db/collections/outputs/indexer/mod.rs:83]: Those ids are unique across NFTs, Aliases and Foundries ... so querying for a specific id will return the right output type (as long as we stored/computed the explicit id correctly), right?

File: Cargo.toml (hunk 1 above)
Comment (Alexandcoats) [reply Cargo.toml:135]: The other option is to create/find an implementation of blake2b that we can use in a mongo script.

File: src/migrations/1.0.0-beta.31.rs (hunk 3 above)
Comment (Alexandcoats) [reply src/migrations/1.0.0-beta.31.rs:77]: Again, `updateMany` cannot work that way. It can only update multiple items that all have the *same* (i.e. non-unique) key.

File: src/migrations/1.0.0-beta.31.rs (hunk 4 above)
Comment (Alexandcoats) [reply src/migrations/1.0.0-beta.31.rs:55]: Ah, good point. Don't know what I was thinking.

File: src/db/collections/outputs/indexer/mod.rs (hunk 5 above)
Comment (Alexandcoats) [reply src/db/collections/outputs/indexer/mod.rs:83]: That, and also we check the `output.kind` in our queries anyway.

File: src/migrations/1.0.0-beta.31.rs (hunk 6 above)
Comment (Alexandcoats) [reply src/migrations/1.0.0-beta.31.rs:34]: Good point, though it seems this will move anyway.

File: Cargo.toml (hunk 1 above)
Comment (grtlr) [reply Cargo.toml:135]: @Alex6323 found a JavaScript one at some point. But I think we should just use our in-memory implementation.

File: src/migrations/1.0.0-beta.31.rs (hunk 3 above)
Comment (Alexandcoats) [reply src/migrations/1.0.0-beta.31.rs]: Nevermind, I misread your initial comment.

File: src/db/collections/outputs/indexer/mod.rs (hunk 5 above)
Comment (grtlr) [reply src/db/collections/outputs/indexer/mod.rs:83]: I'm talking about the `indexed_output_by_id` query in particular. If a user requests an alias output by using `/api/indexer/v1/outputs/alias/{aliasId}` but for some reason provides an `foundryId`, the response should be either `400` or `404`.

Comment (Alex6323) [reply src/db/collections/outputs/indexer/mod.rs:83]: you're right... I suggest adding:
```rust
impl IndexedId {
    fn kind(&self) -> &str {
        match self {
            Self::Alias(_) => AliasOutput::KIND,
            Self::Foundry(_) => FoundryOutput::KIND,
            Self::Nft(_) => NftOutput::KIND,
        }
    }
}
```
to further restrict the `$match`.

**Review by grtlr — APPROVED** [review]
THanks for adding the additional changes. LGTM, but still needs testing.

File: src/bin/inx-chronicle/migrations/migrate_1_0_0_beta_31.rs [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+                ] } },
+                doc! { "$project": {
+                    "output_id": "$_id"
+                } },
+            ],
+            None,
+        )
+        .await?
+        .map_ok(|res| res.output_id)
+        .try_collect::<Vec<_>>()
```
Comment (Alex6323) [inline src/bin/inx-chronicle/migrations/migrate_1_0_0_beta_31.rs:38]: We probably shouldn't collect those into memory but rather stream them, I guess.

File: src/bin/inx-chronicle/migrations/migrate_1_0_0_beta_31.rs [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+use chronicle::{
+    db::{collections::OutputCollection, MongoDb, MongoDbCollection, MongoDbCollectionExt},
+    types::stardust::block::output::{AliasId, NftId, OutputId},
+};
+use futures::TryStreamExt;
+use mongodb::{bson::doc, options::IndexOptions, IndexModel};
+use serde::Deserialize;
+
+pub const PREV_VERSION: &str = "1.0.0-beta.30";
```
Comment (Alex6323) [inline src/bin/inx-chronicle/migrations/migrate_1_0_0_beta_31.rs:12]: Why is this necessary? The version to run the migration from is already in the Cargo.toml manifest, right?

File: src/bin/inx-chronicle/migrations/migrate_1_0_0_beta_31.rs (hunk 7 above)
Comment (grtlr) [reply src/bin/inx-chronicle/migrations/migrate_1_0_0_beta_31.rs:38]: I agree that would be the correct way of doing things but I believe we will get rid of these migrations for a final RC anyways. So I would not worry about it now.

File: src/bin/inx-chronicle/migrations/migrate_1_0_0_beta_31.rs (hunk 8 above)
Comment (Alexandcoats) [reply src/bin/inx-chronicle/migrations/migrate_1_0_0_beta_31.rs:12]: Not if you've updated the application already.

File: src/bin/inx-chronicle/migrations/migrate_1_0_0_beta_31.rs [hunk 9]
```diff
@@ -57,9 +57,9 @@ pub async fn migrate(db: &MongoDb) -> eyre::Result<()> {
                 "output.kind": "alias",
                 "output.alias_id": { "$ne": AliasId::implicit() },
             } },
-            doc! { "$set": {
+            vec![doc! { "$set": {
```
[the remark this answers is not in the document]
Comment (Alexandcoats) [reply src/bin/inx-chronicle/migrations/migrate_1_0_0_beta_31.rs:60]: Unfortunately no 🤷 Apparently the update modification type does not `impl From<[T]>`.

**Review by grtlr — APPROVED** [review]

**Review by Alex6323 — APPROVED** [review]

**Review by grtlr — APPROVED** [review]
